contracts/src/Staking.sol

```diff
@@ -236,51 +245,142 @@ contract Staking is Ownable {
      * @param amount The amount of tokens to stake
      */
     function stake(address validator, uint256 amount) external {
-        if (amount == 0) revert InvalidAmount();
-        if (validator == address(0)) revert InvalidAddress();
-        if (!isValidator[validator]) revert NotValidator();
+        require(amount != 0, InvalidAmount());
+        require(validator != address(0), InvalidAddress());
+        require(isValidator[validator], NotValidator());
 
         stakes[msg.sender][validator] += amount;
         totalStakes[validator] += amount;
         totalStakedAmount += amount;
         emit StakeIncreased(msg.sender, validator, amount);
 
-        safeToken.safeTransferFrom(msg.sender, address(this), amount);
+        SAFE_TOKEN.safeTransferFrom(msg.sender, address(this), amount);
     }
 
     /*
      * @notice Initiate a withdrawal of staked tokens
      * @param validator The validator address to withdraw from
      * @param amount The amount of tokens to withdraw
+     * @dev This function adds the withdrawal to a time-ordered linked list (by claimableAt)
+     *      This should be the function used in the interface. There is another function that allows
+     *      specifying the position in the queue, but it is intended for usage only if absolutely necessary.
      */
     function initiateWithdrawal(address validator, uint256 amount) external {
-        if (amount == 0) revert InvalidAmount();
-        if (stakes[msg.sender][validator] < amount) revert InsufficientStake();
+        require(amount != 0, InvalidAmount());
+        require(stakes[msg.sender][validator] >= amount, InsufficientStake());
 
         stakes[msg.sender][validator] -= amount;
         totalStakes[validator] -= amount;
         totalStakedAmount -= amount;
         totalPendingWithdrawals += amount;
 
-        // Calculate claimable timestamp
+        // Calculating & casting claimable timestamp
         uint128 claimableAt = uint128(block.timestamp + withdrawDelay);
 
         // Generate new withdrawal ID and create node
-        uint128 withdrawalId = nextWithdrawalId++;
-        withdrawalNodes[withdrawalId] = WithdrawalNode({amount: amount, claimableAt: claimableAt, next: 0});
+        uint64 withdrawalId = nextWithdrawalId++;
+        withdrawalNodes[withdrawalId] = WithdrawalNode({amount: amount, claimableAt: claimableAt, previous: 0, next: 0});
 
         // Add to queue
         WithdrawalQueue storage queue = withdrawalQueues[msg.sender][validator];
+        // If queue is empty, set head and tail to new node
         if (queue.head == 0) {
-            // Queue is empty
-            withdrawalQueues[msg.sender][validator] = WithdrawalQueue({head: withdrawalId, tail: withdrawalId});
-        } else {
-            // Link from tail
-            withdrawalNodes[queue.tail].next = withdrawalId;
+            queue.head = withdrawalId;
             queue.tail = withdrawalId;
```

Comment:
Not for this PR, but it would be worth checking if this is a single or multiple `SSTORE`s.

Optimizations aren't important ATM, so I'm happy if this gets dumped into a "things to optimise later" GitHub issue.

Reply:
Updated 👍🏾 I think it might be multiple `SSTORE`.